# Data Cleaning y Procesamiento de Datos

In [ ]:
#Primero visualizamos la información en nuestra base de datos. 
import pandas as pd
import numpy as np

df=pd.read_csv("datos_full.csv")
df.tail() #Con esto encontramos que hay varias columnas con NA

fecha              delito   x   y        colonia  \
210701  2023-01-25  Violencia familiar NaN NaN  NO DISPONIBLE   
210702  2023-02-21  Violencia familiar NaN NaN  NO DISPONIBLE   
210703  2023-03-09  Violencia familiar NaN NaN  NO DISPONIBLE   
210704  2023-02-10  Violencia familiar NaN NaN  NO DISPONIBLE   
210705  2023-01-03  Violencia familiar NaN NaN  NO DISPONIBLE   

                   municipio  clave_mun   hora bien_afectado zona_geografica  
210701                  TALA         83  12:30    La familia        Interior  
210702  AHUALULCO DE MERCADO          3  10:00    La familia        Interior  
210703                  TALA         83   N.D.    La familia        Interior  
210704               BOLAÑOS         19  00:00    La familia        Interior  
210705             CIHUATLAN         22  20:00    La familia        Interior

In [ ]:
#Para nuestro objetivo, no necesitamos los puntos de datos con NA, por eso los descartamos:
df = df.dropna()
#visualizamos de nueva cuenta
df.tail()

fecha              delito           x          y        colonia  \
210173  2023-02-01  Violencia familiar -104.570231  19.246050    LOMA BONITA   
210174  2023-02-03  Violencia familiar -104.485485  19.551139  LOS TECOMATES   
210175  2023-02-06  Violencia familiar -104.565752  19.238043         CENTRO   
210176  2023-02-07  Violencia familiar -104.787389  19.277587  LA MANZANILLA   
210177  2023-01-04  Violencia familiar -104.685846  19.234079     SANTA ROSA   

                municipio  clave_mun   hora bien_afectado zona_geografica  
210173          CIHUATLAN         22  12:00    La familia        Interior  
210174  CASIMIRO CASTILLO         21  01:00    La familia        Interior  
210175          CIHUATLAN         22  02:00    La familia        Interior  
210176          LA HUERTA         43  15:00    La familia        Interior  
210177          CIHUATLAN         22  06:20    La familia        Interior

Con lo anterior, pasamos de 210705 datos a 210177 de datos 'limpios'

In [ ]:
#Visualizamos las diferentes columnas
df.columns

Index(['fecha', 'delito', 'x', 'y', 'colonia', 'municipio', 'clave_mun',
       'hora', 'bien_afectado', 'zona_geografica'],
      dtype='object')

In [ ]:
#Nos enfocaremos únicamente en los ciertos valores:

#zona_geografica = AMG
df = df[df['zona_geografica'] == 'AMG']

In [ ]:
#delito = Robo a persona
df = df[df['delito'] == 'Robo a persona']

In [ ]:
#visualizamos nuestro nuevo set de datos:
df.head()

fecha          delito           x          y  \
25116  2020-01-05  Robo a persona -103.388453  20.640806   
25117  2020-02-04  Robo a persona -103.394789  20.643607   
25118  2020-02-04  Robo a persona -103.387858  20.674246   
25119  2020-02-04  Robo a persona -103.344866  20.675268   
25120  2020-01-05  Robo a persona -103.319531  20.696336   

                        colonia    municipio  clave_mun   hora  bien_afectado  \
25116                   LA CRUZ  GUADALAJARA         39  20:00  El patrimonio   
25117    BOSQUES DE LA VICTORIA  GUADALAJARA         39   N.D.  El patrimonio   
25118         VALLARTA PONIENTE  GUADALAJARA         39  19:00  El patrimonio   
25119               ZONA CENTRO  GUADALAJARA         39  00:00  El patrimonio   
25120  CIRCUNVALACION BELISARIO  GUADALAJARA         39  15:00  El patrimonio   

      zona_geografica  
25116             AMG  
25117             AMG  
25118             AMG  
25119             AMG  
25120             AMG

Ahora contamos con 25120 filas de eventos delictivos bajo la categoria de 'Robo a persona'

Nuestro algoritmo de clustering espectral, solo tomará valores de latitud y longitud, pues usaremos HDBSCAN y Distancia de Haversine. 

In [ ]:
df = df[['x','y']]
# y también, necesitamos los datos en radianes, para lo que primero tenemos que convertir nuestros datos en un array 'X'
X = np.array(df)

In [ ]:
#A continuación, convertir los datos de latitud y longitud a radianes.
X_scaled = np.radians(X)
# visualizamos
X_scaled

array([[-1.80446891,  0.36025003],
       [-1.8045795 ,  0.36029892],
       [-1.80445852,  0.36083367],
       ...,
       [-1.80262878,  0.35887622],
       [-1.80257014,  0.35883076],
       [-1.80344141,  0.35899305]])